### PR TITLE
Add Refresh Creds And Match CWA Creds Order

### PR DESCRIPTION
**Pull request by sethAmazon**
**Description:** 
Refresh creds do not work in skd by default 

**Link to tracking Issue:** 
N/A

**Testing:** 
Copied refresh creds test from cwa

**Documentation:** 
N/A

### CODE REVIEW

**Review by okankoAMZ — APPROVED** [review]

File: internal/aws/awsutil/conn.go [hunk 1]
```diff
@@ -203,29 +203,51 @@ func GetAWSConfigSession(logger *zap.Logger, cn ConnAttr, cfg *AWSSessionSetting
...
+// order
+// 1. shared profile
+// 2. env var
+// 3. shared creds file
+// 4. ec2 creds
+// 5. override creds providers
+// explicit keys do not make sense in this context since contrib does not take in explicit keys
+// and cwa does not provide explicit keys to config but come from these methods of getting creds
+// @TODO make this more upstream friendly working in the future
+func customCredentialProvider(cfg *AWSSessionSettings, config *aws.Config, sess *session.Session) []credentials.Provider {
```
Comment (sky333999) [inline internal/aws/awsutil/conn.go:224]: Do we need to pass in `sess`? I dont see it being used anywhere.

File: internal/aws/awsutil/conn.go [hunk 2]
```diff
@@ -357,6 +390,20 @@ func getSTSCredsFromPrimaryRegionEndpoint(logger *zap.Logger, t *session.Session
 	return nil
 }
 
+func GetRoleCredsPrimaryRegionEndpoint(t *session.Session, roleArn string, region string) credentials.Provider {
```
Comment (sky333999) [inline internal/aws/awsutil/conn.go:393]: Where is this being used?

File: internal/aws/awsutil/conn.go [hunk 3]
```diff
@@ -203,29 +203,51 @@ func GetAWSConfigSession(logger *zap.Logger, cn ConnAttr, cfg *AWSSessionSetting
...
 	} else if cfg.Profile != "" {
-		credProviders = append(credProviders, &credentials.SharedCredentialsProvider{Filename: "", Profile: cfg.Profile})
+		sharedCredentialsProvider := credentials.SharedCredentialsProvider{Filename: "", Profile: cfg.Profile}
+		credProviders = append(credProviders, &sharedCredentialsProvider)
+		credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{sharedCredentialsProvider: &sharedCredentialsProvider})
 	}
-	credProviders = append(credProviders, defaultCredProviders...)
+	credProviders = append(credProviders, &credentials.EnvProvider{})
+	if cfg.Profile == "" && cfg.SharedCredentialsFile != nil && len(cfg.SharedCredentialsFile) > 0 {
+		for _, file := range cfg.SharedCredentialsFile {
```
Comment (sky333999) [inline internal/aws/awsutil/conn.go:241]: Lines 241-244 seem to be the same as lines 229-233 so I initially thought we could drop the `cfg.Profile != ""` check on line 228 and thatll reduce the code, but doing that would change the creds order wouldnt it?

File: internal/aws/awsutil/conn.go [hunk 4]
```diff
@@ -203,29 +203,51 @@ func GetAWSConfigSession(logger *zap.Logger, cn ConnAttr, cfg *AWSSessionSetting
...
+			sharedCredentialsProvider := credentials.SharedCredentialsProvider{Filename: file, Profile: cfg.Profile}
+			credProviders = append(credProviders, &sharedCredentialsProvider)
+			credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{sharedCredentialsProvider: &sharedCredentialsProvider})
+		}
+	}
+	sharedCredentialsProvider := credentials.SharedCredentialsProvider{Filename: "", Profile: ""}
```
Comment (sky333999) [inline internal/aws/awsutil/conn.go:247]: What is this for? Is this for default profile?

Comment (sky333999) [reply internal/aws/awsutil/conn.go:247]: Looks like the default profile we use for onprem is diff, `AmazonCloudWatchAgent`: https://github.com/aws/amazon-cloudwatch-agent/blob/46146905924b34093de4dfda572372f7621c560a/translator/util/sdkutil.go#L170

File: internal/aws/awsutil/conn.go (hunk 3 above)
Comment (zhihonl) [reply internal/aws/awsutil/conn.go:241]: Confused on this as well. Do we need the `cfg.Profile != ""` check? It looks empty or non-empty, we are setting the credential to whatever is in the file, and if empty, SDK will just fallback to [default](https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/#SharedCredentialsProvider)?

Comment (sethAmazon) [reply internal/aws/awsutil/conn.go:241]: This is because of the creds order. We want when a profile is given to always be before a profile not being given in the cred provider. This is based on the comments. Take for instance we pass in a profile but no file.

File: internal/aws/awsutil/conn.go (hunk 4 above)
Comment (sethAmazon) [reply internal/aws/awsutil/conn.go:247]: Yes this is the default

**Review by sky333999 — DISMISSED** [review]

File: internal/aws/awsutil/conn.go [hunk 5]
```diff
@@ -203,29 +203,58 @@ func GetAWSConfigSession(logger *zap.Logger, cn ConnAttr, cfg *AWSSessionSetting
...
-	defaultCredProviders := defaults.CredProviders(config, defaults.Handlers())
 	overrideCredProviders := override.GetCredentialsChainOverride().GetCredentialsChain()
 	credProviders := make([]credentials.Provider, 0)
-	// if is for differently configured shared creds file location
-	// else if is for diff profile but no change in creds file ex run in containers
-	if cfg.SharedCredentialsFile != nil && len(cfg.SharedCredentialsFile) > 0 {
+	// DO NOT TRY TO OPTIMIZE THIS CODE !!!
+	// If you do please do not change order
+	// this is to keep ordering where if there is a profile we want to try those creds first
+	if cfg.Profile != "" && cfg.SharedCredentialsFile != nil && len(cfg.SharedCredentialsFile) > 0 {
```
Comment (jefchien) [inline internal/aws/awsutil/conn.go:224]: nit: Don't need to do a nil check if you're doing a len check. len(nil) is 0.

File: internal/aws/awsutil/conn.go [hunk 6]
```diff
@@ -203,29 +203,58 @@ func GetAWSConfigSession(logger *zap.Logger, cn ConnAttr, cfg *AWSSessionSetting
...
+// order
+// 1. shared profile
+// 2. env var
+// 3. shared creds file
+// 4. ec2 creds
+// 5. override creds providers
+// explicit keys do not make sense in this context since contrib does not take in explicit keys
+// and cwa does not provide explicit keys to config but come from these methods of getting creds
+// @TODO make this more upstream friendly working in the future
 func customCredentialProvider(cfg *AWSSessionSettings, config *aws.Config) []credentials.Provider {
```
Comment (jefchien) [inline internal/aws/awsutil/conn.go:218]: To recap the discussion we had earlier, this should only be the custom credential chain that'll sit before the SDK credential chain. The override provider should be appended to the front of the chain. I still think we can follow the agent and have the roleARN STS provider be an alternative chain to the refreshable shared credential providers. https://github.com/aws/amazon-cloudwatch-agent/blob/main/cfg/aws/credentials.go#L123

File: internal/aws/awsutil/conn.go [hunk 7]
```diff
@@ -267,24 +296,16 @@ func ProxyServerTransport(logger *zap.Logger, config *AWSSessionSettings) (*http
 	return transport, nil
 }
 
-func (c *Conn) newAWSSession(logger *zap.Logger, roleArn string, region string, profile string, sharedCredentialsFile []string) (*session.Session, error) {
+func (c *Conn) newAWSSession(logger *zap.Logger, cfg *AWSSessionSettings, region string) (*session.Session, error) {
```
Comment (jefchien) [inline internal/aws/awsutil/conn.go:299]: If we move the STS credential provider to the credential chain, then we can pass in the `*credential.Credentials` that we get from that chain instead of the `*AWSSessionSettings`.

File: internal/aws/awsutil/conn.go [hunk 8]
```diff
@@ -203,29 +203,58 @@ func GetAWSConfigSession(logger *zap.Logger, cn ConnAttr, cfg *AWSSessionSetting
...
+			})
+		}
 	}
-	credProviders = append(credProviders, defaultCredProviders...)
+	credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{
+		sharedCredentialsProvider: &credentials.SharedCredentialsProvider{
+			Filename: "", Profile: "",
+		},
+	})
+	credProviders = append(credProviders, defaults.RemoteCredProvider(*config, defaults.Handlers()))
```
Comment (jefchien) [inline internal/aws/awsutil/conn.go:257]: If we rely on the SDK ordering as a fallback, then we don't need these or the envProvider in our chain.

File: internal/aws/awsutil/conn.go [hunk 9]
```diff
@@ -383,3 +356,51 @@ func getPartition(region string) string {
...
+	credProviders := make([]credentials.Provider, 0)
+	for _, provider := range overrideCredProviders {
+		for _, file := range cfg.SharedCredentialsFile {
+			credProviders = append(credProviders, provider(file))
+		}
+	}
+	// DO NOT TRY TO OPTIMIZE THIS CODE !!!
+	// If you do please do not change order
+	// this is to keep ordering where if there is a profile we want to try those creds first
+	if cfg.Profile != "" && cfg.SharedCredentialsFile != nil {
```
Comment (jefchien) [inline internal/aws/awsutil/conn.go:378]: If you remove the `cfg.Profile` check, then we don't need the `if cfg.Profile == "" && cfg.SharedCredentialsFile != nil` block.

File: internal/aws/awsutil/conn.go [hunk 10]
```diff
@@ -383,3 +356,51 @@ func getPartition(region string) string {
...
+// 2. shared creds file
+// 3. shared profile
+// explicit keys do not make sense in this context since contrib does not take in explicit keys
+// and cwa does not provide explicit keys to config but come from these methods of getting creds
+// @TODO make this more upstream friendly working in the future
+func customCredentialProvider(cfg *AWSSessionSettings) []credentials.Provider {
+	overrideCredProviders := override.GetCredentialsChainOverride().GetCredentialsChain()
+	credProviders := make([]credentials.Provider, 0)
+	for _, provider := range overrideCredProviders {
+		for _, file := range cfg.SharedCredentialsFile {
```
Comment (jefchien) [inline internal/aws/awsutil/conn.go:371]: This is throwing me off. Why are we creating an override credential provider for each of the shared credentials files?

File: internal/aws/awsutil/refreshable_shared_credentials_provider.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+const (
+	defaultExpiryWindow = time.Minute * 10
+)
+
+type RefreshableSharedCredentialsProvider struct {
+	credentials.Expiry
+	sharedCredentialsProvider *credentials.SharedCredentialsProvider
+
+	// Retrival frequency, if the value is 15 minutes, the credentials will be retrieved every 15 minutes.
+	ExpiryWindow time.Duration
```
Comment (jefchien) [inline internal/aws/awsutil/refreshable_shared_credentials_provider.go:33]: nit: Why do we have this if it's never set?

File: internal/aws/awsutil/conn.go [hunk 12]
```diff
@@ -383,3 +356,51 @@ func getPartition(region string) string {
...
 }
+
+// order
+// 1. override creds providers
+// 2. shared creds file
+// 3. shared profile
+// explicit keys do not make sense in this context since contrib does not take in explicit keys
+// and cwa does not provide explicit keys to config but come from these methods of getting creds
+// @TODO make this more upstream friendly working in the future
+func customCredentialProvider(cfg *AWSSessionSettings) []credentials.Provider {
```
Comment (jefchien) [inline internal/aws/awsutil/conn.go:367]: nit: This is a credential provider chain rather than a single custom credential provider. 

```suggestion
func getCredentialProviderChain(cfg *AWSSessionSettings) []credentials.Provider {
```

File: internal/aws/awsutil/conn.go [hunk 13]
```diff
@@ -369,8 +332,18 @@ func getSTSRegionalEndpoint(r string) string {
 	return e
 }
 
-func GetDefaultSession(logger *zap.Logger, options session.Options) (*session.Session, error) {
-	result, serr := session.NewSessionWithOptions(options)
+func GetDefaultSession(logger *zap.Logger, cfg *AWSSessionSettings) (*session.Session, error) {
+	awsConfig := &aws.Config{}
+	cCredentialProvider := customCredentialProvider(cfg)
+	for i := 0; i < len(cCredentialProvider); i++ {
+		retrieve, err := cCredentialProvider[i].Retrieve()
```
Comment (jefchien) [inline internal/aws/awsutil/conn.go:339]: This is a bit different from the CWA implementation. In CWA, we don't retrieve the actual credentials when evaluating the chain. All it takes for a credential provider to get selected is for the condition check to pass. https://github.com/aws/amazon-cloudwatch-agent/blob/main/cfg/aws/credentials.go#L61

File: internal/aws/awsutil/refreshable_shared_credentials_provider.go [hunk 14]
```diff
@@ -0,0 +1,48 @@
+// Copyright The OpenTelemetry Authors
+// Portions of this file Copyright 2018-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (jefchien) [inline internal/aws/awsutil/refreshable_shared_credentials_provider.go:2]: nit: This header seems wrong or outdated.

File: internal/aws/awsutil/conn.go (hunk 7 above)
Comment (sethAmazon) [reply internal/aws/awsutil/conn.go:241]: Role needs to be separated. As in if you give a role you do not use other creds.

File: internal/aws/awsutil/conn.go (hunk 10 above)
Comment (sethAmazon) [reply internal/aws/awsutil/conn.go:371]: Because we pass in odin path through this field. That is how we do it in the current agent. Take a look at how the java agent injects odin.

File: internal/aws/awsutil/refreshable_shared_credentials_provider.go (hunk 11 above)
Comment (sethAmazon) [reply internal/aws/awsutil/refreshable_shared_credentials_provider.go:33]: So we can set it in the test.

File: internal/aws/awsutil/conn.go (hunk 13 above)
Comment (sky333999) [reply internal/aws/awsutil/conn.go:339]: True. But does doing it this way hurt? Instead of just checking for file or profile existence, I guess we try to validate if they can actually be used to retrieve creds.

Comment (jefchien) [reply internal/aws/awsutil/conn.go:339]: Not necessarily, but it can lead to different behavior. The original implementation meant that if we had a `Profile` or `Filename` set, then it would only try the SharedCredentialsProvider even if the properties were configured incorrectly. Now, we'll fallback on the SDK chain.

File: internal/aws/awsutil/conn.go (hunk 7 above)
Comment (sky333999) [reply internal/aws/awsutil/conn.go:241]: I think Jeffrey is saying add it to our custom order so it still takes precedence over the others (itll do it again as part of the fallback logic, but that might be ok?)

File: internal/aws/awsutil/conn.go [hunk 15]
```diff
@@ -383,3 +356,51 @@ func getPartition(region string) string {
...
+		}
+	} else if cfg.Profile != "" {
+		credProviders = append(credProviders, &RefreshableSharedCredentialsProvider{
+			sharedCredentialsProvider: &credentials.SharedCredentialsProvider{
+				Filename: "",
+				Profile:  cfg.Profile,
+			},
+		})
+	}
+	if cfg.Profile == "" && cfg.SharedCredentialsFile != nil {
```
Comment (sky333999) [inline internal/aws/awsutil/conn.go:395]: nit: should this be `else if`?

Im ok with it for now, but this code will be confusing when I look at it in the future.
Wonder if it can be simplified to just create two vars for profile and files. files can be initialized using a for loop when cfg.SharedCredentialsFile is not nil and initialized to a slice with "" when it is nil.
Looks like we always set cfg.Profile even if its "".

**Comment by sethAmazon** [comment]
Interesting that you can iterate a nil and len a nil in golang without kernel panic.

File: internal/aws/awsutil/conn.go (hunk 7 above)
Comment (jefchien) [reply internal/aws/awsutil/conn.go:241]: I originally thought it could be an early return in the `getCredentialProviderChain` where it just returns a slice with the STS provider so we could have all our credential chain in one place, but looking at it again, that would break it. The STS provider requires a default session to get the role credentials in the first place so we'd end up in a circular dependency if the roleARN was present.

File: internal/aws/awsutil/conn.go (hunk 13 above)
Comment (sethAmazon) [reply internal/aws/awsutil/conn.go:339]: I do not think that behavior can be upstreamed.

Comment (sethAmazon) [reply internal/aws/awsutil/conn.go:339]: I think falling back if it can't get creds is fine even if the current agent does not.

File: internal/aws/awsutil/conn.go [hunk 16]
```diff
@@ -369,17 +371,120 @@ func getSTSRegionalEndpoint(r string) string {
...
+	case dcaPartition:
+		return dcaFallbackRegion
+	case lckPartition:
+		return lckFallbackRegion
+	default:
+		return classicFallbackRegion
+	}
+}
+
+func getEndpointPartition(region string) endpoints.Partition {
```
Comment (jefchien) [inline internal/aws/awsutil/conn.go:481]: nit: `getPartition` can probably be removed and replaced with this one.

**Review by jefchien — DISMISSED** [review]

**Review by sky333999 — APPROVED** [review]

File: internal/aws/awsutil/conn.go (hunk 16 above)
Comment (sethAmazon) [reply internal/aws/awsutil/conn.go:490]: I'm just going to keep the current contrib code for now and decide what to do later.

File: internal/aws/awsutil/conn.go [hunk 17]
```diff
@@ -301,13 +308,13 @@ func (c *Conn) newAWSSession(logger *zap.Logger, roleArn string, region string,
 // getSTSCreds gets STS credentials from regional endpoint. ErrCodeRegionDisabledException is received if the
 // STS regional endpoint is disabled. In this case STS credentials are fetched from STS primary regional endpoint
 // in the respective AWS partition.
-func getSTSCreds(logger *zap.Logger, region string, roleArn string) (*credentials.Credentials, error) {
-	t, err := GetDefaultSession(logger, session.Options{})
+func getSTSCreds(logger *zap.Logger, region string, cfg *AWSSessionSettings) (*credentials.Credentials, error) {
```
Comment (jefchien) [inline internal/aws/awsutil/conn.go:311]: Can we revert the changes to `getSTSCreds` since we're no longer using it?

Comment (sethAmazon) [reply internal/aws/awsutil/conn.go:311]: I made a change to GetDefaultSession

**Review by jefchien — APPROVED** [review]

**Review by sky333999 — APPROVED** [review]
